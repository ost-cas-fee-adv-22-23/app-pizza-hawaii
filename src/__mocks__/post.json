{
	"id": "01GZFZES838XW8SV8QVWCJSN6C",
	"creator": "195305735549092097",
	"text": "<p>Hier ist noch ein <a href=\"/tag/hashtag\">#hashtag</a></p>",
	"mediaUrl": null,
	"mediaType": null,
	"likeCount": 0,
	"likedByUser": false,
	"type": "post",
	"replyCount": 3,
	"createdAt": "2023-05-03T04:55:45.155Z",
	"user": {
		"posterImage": "https://picsum.photos/seed/1953057355490920971/1466/1060/",
		"bio": "Hello my name is Peter!. I am a Gamer. Troublemaker. Hardcore food buff. Hipster-friendly creator. Falls down a lot. Alcohol geek. Organizer.",
		"createdAt": "2023-10-24T22:00:00.000Z",
		"city": "Lyss",
		"displayName": "Peter Manser",
		"id": "195305735549092097",
		"userName": "peter",
		"firstName": "Peter",
		"lastName": "Manser",
		"avatarUrl": "https://cas-fee-advanced-ocvdad.zitadel.cloud/assets/v1/179828644300980481/users/195305735549092097/avatar?v=3b1662b598359b0221e987ebc08375f7",
		"profileLink": "/user/195305735549092097"
	},
	"replies": [
		{
			"id": "01GZZWQW4XP9M7TW89X5ERQ84F",
			"creator": "195305735549092097",
			"text": "<p>2</p>",
			"mediaUrl": null,
			"mediaType": null,
			"likeCount": 1,
			"likedByUser": false,
			"type": "reply",
			"parentId": "01GZFZES838XW8SV8QVWCJSN6C",
			"createdAt": "2023-05-09T09:16:08.221Z",
			"user": {
				"posterImage": "https://picsum.photos/seed/1953057355490920971/1466/1060/",
				"bio": "Hello my name is Peter!. I am a Gamer. Troublemaker. Hardcore food buff. Hipster-friendly creator. Falls down a lot. Alcohol geek. Organizer.",
				"createdAt": "2023-10-24T22:00:00.000Z",
				"city": "Lyss",
				"displayName": "Peter! Manser",
				"id": "195305735549092097",
				"userName": "peter",
				"firstName": "Peter!",
				"lastName": "Manser",
				"avatarUrl": "https://cas-fee-advanced-ocvdad.zitadel.cloud/assets/v1/179828644300980481/users/195305735549092097/avatar?v=3b1662b598359b0221e987ebc08375f7",
				"profileLink": "/user/195305735549092097"
			}
		},
		{
			"id": "01GZZWQH1HCEP4H7R9EQEY1E25",
			"creator": "195305735549092097",
			"text": "<p>1</p>",
			"mediaUrl": null,
			"mediaType": null,
			"likeCount": 1,
			"likedByUser": false,
			"type": "reply",
			"parentId": "01GZFZES838XW8SV8QVWCJSN6C",
			"createdAt": "2023-05-09T09:15:56.849Z",
			"user": {
				"posterImage": "https://picsum.photos/seed/1953057355490920971/1466/1060/",
				"bio": "Hello my name is Peter!. I am a Gamer. Troublemaker. Hardcore food buff. Hipster-friendly creator. Falls down a lot. Alcohol geek. Organizer.",
				"createdAt": "2023-10-24T22:00:00.000Z",
				"city": "Lyss",
				"displayName": "Peter! Manser",
				"id": "195305735549092097",
				"userName": "peter",
				"firstName": "Peter!",
				"lastName": "Manser",
				"avatarUrl": "https://cas-fee-advanced-ocvdad.zitadel.cloud/assets/v1/179828644300980481/users/195305735549092097/avatar?v=3b1662b598359b0221e987ebc08375f7",
				"profileLink": "/user/195305735549092097"
			}
		}
	]
}
