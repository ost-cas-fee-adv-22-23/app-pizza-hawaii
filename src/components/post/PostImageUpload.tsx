import { Icon, Label } from '@smartive-education/pizza-hawaii';
import React, { FC } from 'react';
import { useDropzone } from 'react-dropzone';

type TPostImageUpload = {
	onNewFile: (file: File) => void;
};

export const PostImageUpload: FC<TPostImageUpload> = ({ onNewFile }) => {
	const handleFileDrop = (acceptedFiles: File[]) => {
		const uploadedFile = acceptedFiles?.[0];
		onNewFile && onNewFile(uploadedFile);
	};

	const { getRootProps, getInputProps } = useDropzone({
		accept: {
			'image/png': [],
			'image/jpeg': [],
			'image/jpg': [],
			'image/gif': [],
		},
		maxSize: 5000000,
		onDrop: handleFileDrop,
		onDropRejected: (rejectedFiles, error) => {
			console.error('onDropRejected:', rejectedFiles, error);
		},
		onError: (error) => {
			console.error('onError:', error);
		},
	});
	// eslint-disable-next-line no-console
	console.log('next release will have a file upload');
	return (
		<div className="p-2 h-48 cursor-pointer flex justify-center align-middle bg-slate-100 text-slate-600">
			<div {...getRootProps({ className: 'dropzone' })}>
				<input {...getInputProps()} placeholder="Um Datei uploaden hierhinziehen oder clicken" />
				<div className="block justify-center text-center align-middle p-10">
					<Icon size="L" name="upload" />
					<br />
					<Label as="span" size="M">
						Datei hier hineinziehen oder clicken
					</Label>
					<br />
					<Label as="span" size="S">
						JPG, GIF oder PNG, maximal 5 MB
					</Label>
				</div>
			</div>
		</div>
	);
};
