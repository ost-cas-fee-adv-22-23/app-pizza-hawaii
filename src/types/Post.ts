import { TUser } from './User';

export type TPost = {
	id: string;
	creator: TUser | string;
	text: string;
	mediaUrl?: string | null;
	mediaType: 'image/jpeg';
	likeCount: number;
	likedByUser: boolean | string;
	replyCount?: number;
	createdAt: string;
	type: 'post' | 'reply' | 'deleted';
	replies?: TPost[];
};
