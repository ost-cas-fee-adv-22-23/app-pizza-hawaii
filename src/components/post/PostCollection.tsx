import { Button, Grid, Headline } from '@smartive-education/pizza-hawaii';
import { useSession } from 'next-auth/react';
import { FC, useEffect, useReducer, useState } from 'react';

import { useActiveTabContext } from '../../context/useActiveTab';
import useIncreasingInterval from '../../hooks/useIncreasingInterval';
import PCReducer, {
	ActionType as PCActionType,
	initialState as initialPCState,
	TActionType,
} from '../../reducer/postCollectionReducer';
import { services } from '../../services';
import { TPost } from '../../types';
import { default as ULID } from '../../utils/UlidHelper';
import { PostList } from '../post/PostList';
import { PostCreator, TAddPostProps } from './PostCreator';

type TPostCollectionFilter = {
	creator?: string;
	likedBy?: string[];
	tags?: string[];
};

export type TPostCollectionProps = {
	headline?: string;
	posts: TPost[];
	canAdd?: boolean;
	canLoadMore: boolean;
	filter?: TPostCollectionFilter;
	autoUpdate?: boolean;
};

enum LoadRequestType {
	LOAD_NOT_NEEDED = 'LOAD_NOT_NEEDED',
	LOAD_ADDED = 'LOAD_ADDED',
	LOAD_UPDATED = 'LOAD_UPDATED',
	LOAD_ADDED_AND_UPDATED = 'LOAD_ADDED_AND_UPDATED',
}

type TLoadRequestType = keyof typeof LoadRequestType;

export const PostCollection: FC<TPostCollectionProps> = ({
	headline,
	posts: initialPosts = [],
	canAdd = false,
	canLoadMore = false,
	autoUpdate,
	filter = {},
}) => {
	const { data: session } = useSession();
	const { isActive: tabIsActive } = useActiveTabContext();

	const [postState, postDispatch] = useReducer(PCReducer, {
		...initialPCState,
		posts: initialPosts,
	});

	/**
	 *
	 * ============== LOAD MORE MECHANISM ==============
	 *
	 */

	const [canLoadmore, setCanLoadmore] = useState(canLoadMore);

	const onLoadmoreBtn = async () => {
		// activate loading state
		postDispatch({ type: PCActionType.LOADING, payload: true });

		// fetch posts older than oldest post
		const { count: olderPostCount, posts: olderPosts } = await services.api.posts.loadmore({
			...filter,
			olderThan: getPostQueryUlids(postState.posts).oldestUlid,
		});

		// append older posts to list
		postDispatch({ type: PCActionType.POSTS_ADD, payload: olderPosts });

		// set canLoadmore to true if there are more posts available
		setCanLoadmore(olderPostCount > 0);

		// deactivate loading state
		postDispatch({ type: PCActionType.LOADING, payload: false });
	};

	/**
	 *
	 * ============== UPDATE MECHANISM ==============
	 *
	 */

	const [loadRequest, setLoadRequest] = useState(LoadRequestType.LOAD_NOT_NEEDED);
	const [updateRequest, setUpdateRequest] = useState<
		| TActionType
		| {
				type: undefined;
		  }
	>({
		type: undefined,
	});

	const onLoadRequestLoad = async (requestObject: TRequestObj) => {
		// fetch posts older than oldest post
		const { posts: loadedPosts } = await services.api.posts.loadmore({
			...filter,
			...requestObject,
		});

		if (!loadedPosts?.length) {
			return {
				posts: [],
				newPosts: [],
				deletedPosts: [],
			};
		}

		// list of posts that have been deleted
		const deletedPostList = postState.posts.filter((post) => {
			return !loadedPosts.find((p) => p.id === post.id);
		});

		// get List of new Posts
		const newPostList = loadedPosts.filter((post) => {
			return !postState.posts.find((p) => p.id === post.id);
		});

		return {
			posts: loadedPosts,
			newPosts: newPostList,
			deletedPosts: deletedPostList,
		};
	};

	useEffect(() => {
		// check not needed because loadRequest is set to LOAD_NOT_NEEDED
		if (loadRequest === LoadRequestType.LOAD_NOT_NEEDED) {
			return;
		}

		// reset loadRequest to LOAD_NOT_NEEDED
		setLoadRequest(LoadRequestType.LOAD_NOT_NEEDED);

		// generate request object from loadRequest and filter
		const requestObject = {
			...filter,
			...getRequestObj(postState.posts, loadRequest),
		};

		(async () => {
			const { posts, newPosts, deletedPosts } = await onLoadRequestLoad(requestObject);

			// if no new or deleted posts are found, do nothing
			if (newPosts.length === 0 && deletedPosts.length === 0) {
				setUpdateRequest({ type: undefined });
				return;
			}

			// define action type for reducer (default: set posts for LOAD_ADDED_AND_UPDATED and LOAD_UPDATED)
			let updateActionType = PCActionType.POSTS_SET;

			// if only new posts are loaded, add them to the existing posts
			if (loadRequest === LoadRequestType.LOAD_ADDED) {
				updateActionType = PCActionType.POSTS_ADD;
			}

			setUpdateRequest({
				payload: posts,
				type: updateActionType,
			});
		})();

		// We only want to trigger this effect when one of the values changes: loadRequest and filter
		// so we disable the exhaustive-deps rule here. Otherwise the effect would be triggered multiple times which is not intended.
		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [loadRequest, filter]);

	const showLatestPosts = () => {
		if (!updateRequest.type) {
			return;
		}

		postDispatch(updateRequest);
		setUpdateRequest({ type: undefined });
	};

	/**
	 * LoadRequest if triggered by interval or by user switching to mumble browser tab
	 */
	useIncreasingInterval(() => {
		// only load new posts if browser tab is active and autoUpdate is enabled
		if (!tabIsActive || !autoUpdate) return;

		// randomizes if load full list (to detect deleted posts) or check just for new posts (2/3 chance)
		// to prevent that all users load the full list at the same time when the interval is triggered
		// full list means to load all already visible posts
		// for our use case this is not necessary, but was fun to implement ;)

		const loadFullList = Math.random() > 0.66;
		setLoadRequest(loadFullList ? LoadRequestType.LOAD_UPDATED : LoadRequestType.LOAD_ADDED);
	});

	useEffect(() => {
		// only load new posts if browser tab is active and autoUpdate is enabled
		if (!tabIsActive || !autoUpdate) return;

		// load full list of posts when user switches to browser tab
		setLoadRequest(LoadRequestType.LOAD_ADDED_AND_UPDATED);
	}, [tabIsActive, autoUpdate]);

	/**
	 *
	 * ============== CRUD FUNCTIONS ==============
	 *
	 */

	const onAddPostFn = async (postData: TAddPostProps): Promise<TPost | null> => {
		const newPost = await services.posts.createPost({
			...postData,
			accessToken: session?.accessToken as string,
		});

		if (!newPost) return null;

		postDispatch({ type: PCActionType.POSTS_ADD, payload: newPost });

		return newPost;
	};

	const onRemovePostFn = async (id: string) => {
		const response = await services.api.posts.remove({ id });

		if (!response.ok) {
			throw new Error('Failed to delete post');
		}
		postDispatch({ type: PCActionType.POSTS_DELETE, payload: id });
	};

	return (
		<>
			{headline && (
				<div className="text-slate-500 mb-8">
					<Headline level={3} as="p">
						{headline}
					</Headline>
				</div>
			)}

			{canAdd && (
				<Grid variant="col" gap="M" marginBelow="M">
					<PostCreator
						textAreaId="post-creator"
						variant="newPost"
						headline="Hey, was geht ab?"
						placeHolderText="Deine Meinung zÃ¤hlt"
						onAddPost={onAddPostFn}
					/>
				</Grid>
			)}
			<div>
				<PostList posts={postState.posts} onRemovePost={onRemovePostFn} showLoadingItems={3} />

				{canLoadmore && (
					<Button colorScheme="slate" onClick={onLoadmoreBtn} disabled={postState.loading}>
						{postState.loading ? 'loading ...' : 'Load more'}
					</Button>
				)}

				{autoUpdate && updateRequest.type && (
					<div className="text-slate-500 mt-8 sticky bottom-4 z-10">
						<Button colorScheme="gradient" size="M" icon="repost" onClick={showLatestPosts}>
							World is changing, update your feed.
						</Button>
					</div>
				)}
			</div>
		</>
	);
};

/**
 *
 * ============== HELPERS ==============
 *
 */

// generate request object for posts
type TRequestObj = {
	limit: number;
	olderThan: string;
	newerThan: string;
};
const getRequestObj = (posts: TPost[], loadRequest: TLoadRequestType) => {
	const { currentUlid, oldestUlid, latestUlid } = getPostQueryUlids(posts);

	// default request object (olderThan: current Date, newerThan: oldest post id - 1, limit: 100) to get all posts
	let requestObject = {
		limit: 100,
	} as TRequestObj;

	switch (loadRequest) {
		case LoadRequestType.LOAD_ADDED:
			// olderThan: current Date (to get latest posts)
			// newerThan: latest post id (to get all new posts)
			requestObject = {
				...requestObject,
				olderThan: currentUlid,
				newerThan: latestUlid,
			};
			break;

		case LoadRequestType.LOAD_UPDATED:
			// olderThan: latest post id + 1 (to get all posts including the latest one)
			// newerThan: oldest post id - 1 (to get all posts including the oldest one)
			requestObject = {
				...requestObject,
				olderThan: ULID.getNewerThan(latestUlid),
				newerThan: ULID.getOlderThan(oldestUlid),
			};
			break;

		case LoadRequestType.LOAD_ADDED_AND_UPDATED:
			// olderThan: current Date (to get latest posts)
			// newerThan: oldest post id - 1 (to get all posts including the oldest one)
			requestObject = {
				...requestObject,
				olderThan: currentUlid,
				newerThan: ULID.getOlderThan(oldestUlid),
			};
			break;

		default:
			break;
	}

	return requestObject;
};

// get oldest post id
const getPostQueryUlids = (posts: TPost[]) => {
	const currentUlid = ULID.getCurrent();

	// check if posts are available
	if (!posts?.length) {
		return {
			currentUlid,
			latestUlid: currentUlid,
			oldestUlid: currentUlid,
		};
	}

	const latestPost = posts[0];
	const oldestPost = posts[posts.length - 1];

	return {
		currentUlid,
		latestUlid: latestPost.id,
		oldestUlid: oldestPost.id,
	};
};
