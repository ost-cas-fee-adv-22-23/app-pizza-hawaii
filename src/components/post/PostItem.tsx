import React, { FC, useState } from 'react';
import NextLink from 'next/link';
import { useSession } from 'next-auth/react';

import {
	Image,
	Label,
	Grid,
	TimeStamp,
	Richtext,
	IconText,
	ImageOverlay,
	CopyToClipboardButton,
	UserContentCard,
	TUserContentCard,
	InteractionButton,
} from '@smartive-education/pizza-hawaii';

import { TPost, TUser } from '../../types';
import ProjectSettings from '../../data/ProjectSettings.json';
import { postsService } from '../../services/api/posts/';
import ImageModal from '../ImageModal';

/*
 * Type
 */

type TPostItemProps = {
	variant: 'detailpage' | 'timeline' | 'response';
	post: TPost;
	onDeletePost?: (id: string) => void;
};

type TPostItemVariantMap = {
	headlineSize: 'S' | 'M' | 'L' | 'XL';
	textSize: 'M' | 'L';
	avatarSize: TUserContentCard['avatarSize'];
	avatarVariant: TUserContentCard['avatarVariant'];
};

/*
 * Style
 */

const postItemVariantMap: Record<TPostItemProps['variant'], TPostItemVariantMap> = {
	detailpage: {
		headlineSize: 'XL',
		textSize: 'L',
		avatarSize: 'M',
		avatarVariant: 'standalone',
	},
	timeline: {
		headlineSize: 'L',
		textSize: 'M',
		avatarSize: 'M',
		avatarVariant: 'standalone',
	},
	response: {
		headlineSize: 'M',
		textSize: 'M',
		avatarSize: 'S',
		avatarVariant: 'subcomponent',
	},
};

export const PostItem: FC<TPostItemProps> = ({ variant, post, onDeletePost }) => {
	const [likedByUser, setLikedByUser] = useState(post.likedByUser);
	const [likeCount, setLikeCount] = useState(post.likeCount);
	const [showImageModal, setShowImageModal] = useState(false);

	const { data: session } = useSession();
	const currentUser = session?.user as TUser;

	const setting = postItemVariantMap[variant] || postItemVariantMap.detailpage;
	const replyCount = post?.replyCount || 0;

	// like and unlike function
	const handleLike = async () => {
		if (likedByUser) {
			postsService.unlike({ id: post.id }).then(() => {
				setLikeCount(likeCount - 1);
			});
		} else {
			postsService.like({ id: post.id }).then(() => {
				setLikeCount(likeCount + 1);
			});
		}
		setLikedByUser(!likedByUser);
	};

	// delete function
	const handleDeletePost = async () => {
		onDeletePost && onDeletePost(post.id);
	};

	// mayby we do a helper function hook or a component for this as ImageModal is used in userpanorama image as well
	// ImageModal function
	const toggleImageModal = () => {
		setShowImageModal(!showImageModal);
	};

	const headerSlotContent = (
		<Grid variant="col" gap="S">
			<Label as="span" size={setting.headlineSize}>
				{`${post.user.displayName}`}
			</Label>
			<Grid variant="row" gap="S">
				<NextLink href={post.user.profileLink}>
					<IconText icon="profile" colorScheme="violet" size="S">
						{post.user.userName}
					</IconText>
				</NextLink>
				<IconText icon="calendar" colorScheme="slate" size="S">
					<TimeStamp date={post.createdAt} />
				</IconText>
			</Grid>
		</Grid>
	);

	return (
		<UserContentCard
			headline={headerSlotContent}
			userProfile={{
				avatar: post.user.avatarUrl,
				userName: post.user.userName,
				href: post.user.profileLink,
			}}
			avatarVariant={setting.avatarVariant}
			avatarSize={setting.avatarSize}
		>
			<Richtext size={setting.textSize}>{post.text}</Richtext>

			{post.mediaUrl && (
				<ImageOverlay preset="enlarge" buttonLabel="Enlarge image in modal" onClick={toggleImageModal}>
					<Image
						width={ProjectSettings.images.post.width}
						height={
							(ProjectSettings.images.post.width / ProjectSettings.images.post.aspectRatio[0]) *
							ProjectSettings.images.post.aspectRatio[1]
						}
						src={post.mediaUrl}
						alt={`Image of ${post.user.displayName}`}
					/>
				</ImageOverlay>
			)}

			<Grid variant="row" gap="M" wrapBelowScreen="md">
				<InteractionButton
					component={NextLink}
					href={`/mumble/${post.id}`}
					isActive={replyCount > 0}
					colorScheme="violet"
					buttonText={replyCount > 0 ? `${replyCount} Comments` : replyCount === 0 ? 'Comment' : '1 Comment'}
					iconName={replyCount > 0 ? 'comment_filled' : 'comment_fillable'}
				/>
				<InteractionButton
					type="button"
					isActive={likeCount > 0}
					colorScheme="pink"
					buttonText={likeCount > 0 ? `${likeCount} Likes` : likeCount === 0 ? 'Like' : '1 Like'}
					iconName={likedByUser ? 'heart_filled' : 'heart_fillable'}
					onClick={handleLike}
				/>

				<CopyToClipboardButton
					defaultButtonText="Copy Link"
					activeButtonText="Link copied"
					shareText={`${process.env.NEXT_PUBLIC_VERCEL_URL}/mumble/${post.id}`}
				/>

				{currentUser && currentUser.id === post.user.id && (
					<InteractionButton
						type="button"
						colorScheme="pink"
						buttonText="Delete"
						iconName="cancel"
						onClick={handleDeletePost}
					/>
				)}
			</Grid>

			{showImageModal && <ImageModal post={post} toggleHandler={setShowImageModal} />}
		</UserContentCard>
	);
};
