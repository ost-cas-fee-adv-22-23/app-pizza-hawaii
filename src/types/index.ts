import { TPost, TRawPost } from './Post';
import { TRawUser, TUser, TUserSimple } from './User';

export type { TUser, TUserSimple, TRawUser, TPost, TRawPost };
