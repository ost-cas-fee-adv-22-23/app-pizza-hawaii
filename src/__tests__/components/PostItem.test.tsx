import { cleanup, render, screen } from '@testing-library/react';
import { useSession } from 'next-auth/react';

import postMock from '../../__mocks__/post.json';
import { PostItem } from '../../components/post/PostItem';
import { TPost } from '../../types/Post';

/**
 * Unit tests for the PostItem component.
 *
 * Basicly we want to know if this often re-used Component is working with the basic funtionailities.
 * We test the following on PostItem
 * 1. Render PostItem component with text
 * 2. Render PostItem component with a link when a hastag is used in the text
 * 3. Render PostItem component with a link when a user is mentioned in the text
 * 4. Render PostItem component with a link when a url is used in the text
 * 5. Render PostItem component with a like button
 * 6. Render PostItem component with a userName with all the correct css classes
 * 7. Render PostItem component with a avatar image with and height 64px
 * 8. Render PostItem component with a copy button for the current user
 * 9. Render PostItem component with a disabled copy button for other users
 * 10. Render PostItem component with a copy button for the current user
 * 11. Render PostItem component with a like button for the current user
 *
 **/

jest.mock('next-auth/react');

export type TPostItemProps = {
	variant: 'detailpage' | 'timeline' | 'response';
	post: TPost;
	onDeletePost?: (id: string) => void;
	onAnswerPost?: (id: string) => void;
};

const propsDetailPage: TPostItemProps = {
	post: postMock as TPost,
	variant: 'detailpage',
	onDeletePost: jest.fn(),
	onAnswerPost: jest.fn(),
};

const propsTimelinePage: TPostItemProps = {
	post: postMock as TPost,
	variant: 'timeline',
	onDeletePost: jest.fn(),
	onAnswerPost: jest.fn(),
};

describe('PostItem renders correctly', () => {
	afterEach(cleanup);
	test('PostItem variant `DetailPage` renders the text of Post', async () => {
		const mockSession = {
			user: { name: 'Filiks Adamski', email: 'filiks.adamski@mumble.com' },
			expires: '1234567890',
		};
		(useSession as jest.Mock).mockReturnValue([mockSession, true]);

		render(<PostItem {...propsDetailPage} />);
		expect(screen.getByText('Hier ist noch ein'));
	});

	test('PostItem variant `DetailPage` renders a link when a hastag is used in the text', async () => {
		const { container } = render(<PostItem {...propsDetailPage} />);
		expect(container.querySelector('#hashtag'));
		expect(container.querySelector('a'));
	});

	test('PostItem variant `DetailPage` renders a link when a user is mentioned in the text', async () => {
		const { container } = render(<PostItem {...propsDetailPage} />);
		expect(container.querySelector('#user'));
		expect(container.querySelector('a'));
	});

	test('PostItem variant `DetailPage` renders a link when a url is used in the text', async () => {
		const { container } = render(<PostItem {...propsDetailPage} />);
		expect(container.querySelector('#url'));
		expect(container.querySelector('a'));
	});
});

/**
 * Test PostItem renders correctly for variant `Timeline` so that the text of Post is rendered correctly
 * with the correct css classes, image size and the correct user name
 */

describe('PostItem renders correctly', () => {
	afterEach(cleanup);

	// test if the text of Post is rendered with the correct css classes
	test('PostItem variant `timeline` displays the User Name with all the correct css classes', async () => {
		render(<PostItem {...propsTimelinePage} />);
		const user = screen.getByText('Peter Manser');
		expect(user).toHaveProperty(
			'className',
			'inline-block font-semibold overflow-hidden text-ellipsis mb-[-0.2em] pb-[0.2em] text-lg'
		);
	});

	// test if the variant `timeline` renders the avatar image with the correct size
	test('PostItem variant `timeline` renders Avatar image with and height 64px', async () => {
		render(<PostItem {...propsTimelinePage} />);
		const avatarImage = screen.getByRole('img');
		expect(avatarImage.getAttribute('width')).toBe('64');
		expect(avatarImage.getAttribute('height')).toBe('64');
	});

	// test if the variant `timeline` renders the text of Post with the correct css classes
	test('displays copy button for the current user', () => {
		const currentUser = {
			id: '1234567890',
			name: 'John Doe',
		};
		useSession.mockReturnValue({ data: { user: currentUser } });

		render(<PostItem variant="detailpage" post={postMock as TPost} />);
		const copyButton = screen.getAllByText('Copy Link');
		expect(copyButton).toHaveLength(1);
		expect(copyButton[0]).toBeInstanceOf(HTMLSpanElement);
	});

	// test if the like button is rendered correctly
	test('PostItem renders the like button correctly', async () => {
		render(<PostItem variant="detailpage" post={postMock as TPost} />);
		const likeButton = screen.getAllByText('Like');
		expect(likeButton).toHaveLength(1);
		expect(likeButton[0]).toBeInstanceOf(HTMLSpanElement);
	});
});
