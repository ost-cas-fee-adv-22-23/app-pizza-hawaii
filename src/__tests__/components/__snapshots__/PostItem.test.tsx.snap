// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostItem renders correctly PostItem variant \`timeline\` renders correctly 1`] = `
<div>
  <div>
    <div
      class="Card relative flex-start justify-center items-start bg-white text-slate-900 rounded-3xl py-l px-xl sm:py-m sm:px-m"
    >
      <div
        class="flex w-full flex-col gap-x-12 gap-y-6 sm:gap-x-8   "
      >
        <div
          class="flex w-full flex-row gap-x-4 gap-y-2 sm:gap-x-2  md:flex-wrap items-center"
        >
          <div
            class="absolute left-0 transform -translate-x-1/2"
          >
            <div
              class="inline-flex rounded-full self-center bg-purple-200 border-6 overflow-hidden h-16 w-16 md:h-12 md:w-12 sm:h-10 sm:w-10"
            >
              <a
                class="transition duration-300 ease-in-out hover:scale-110"
                href="/user/195305735549092097"
              >
                <figure
                  class="grid bg-purple-200 overflow-hidden"
                >
                  <img
                    alt="peter"
                    class="block object-cover h-full w-full"
                    height="64"
                    src="//images.weserv.nl/?url=https%3A%2F%2Fcas-fee-advanced-ocvdad.zitadel.cloud%2Fassets%2Fv1%2F179828644300980481%2Fusers%2F195305735549092097%2Favatar%3Fv%3D3b1662b598359b0221e987ebc08375f7&w=64&fit=cover&h=64"
                    srcset="//images.weserv.nl/?url=https%3A%2F%2Fcas-fee-advanced-ocvdad.zitadel.cloud%2Fassets%2Fv1%2F179828644300980481%2Fusers%2F195305735549092097%2Favatar%3Fv%3D3b1662b598359b0221e987ebc08375f7&w=64&fit=cover&h=64 1x, //images.weserv.nl/?url=https%3A%2F%2Fcas-fee-advanced-ocvdad.zitadel.cloud%2Fassets%2Fv1%2F179828644300980481%2Fusers%2F195305735549092097%2Favatar%3Fv%3D3b1662b598359b0221e987ebc08375f7&w=64&fit=cover&h=64&dpr=2 2x"
                    width="64"
                  />
                </figure>
                <span
                  class="sr-only"
                >
                  Change Avatar
                </span>
              </a>
            </div>
          </div>
          <div
            class="flex w-full flex-col gap-x-4 gap-y-2 sm:gap-x-2   "
          >
            <span
              class="inline-block font-semibold overflow-hidden text-ellipsis mb-[-0.2em] pb-[0.2em] text-lg"
            >
              Peter Manser
            </span>
            <div
              class="flex w-full flex-row gap-x-4 gap-y-2 sm:gap-x-2   "
            >
              <a
                href="/user/195305735549092097"
              >
                <span
                  class="flex items-center leading-none rounded-full overflow-hidden text-ellipsis mb-[-0.2em] pb-[0.2em] text-sm gap-1 px-2 py-1 text-violet-600 hover:text-violet-900 hover:bg-violet-50  group transition-all duration-300"
                >
                  <span
                    class="inline-flex text-[11px] containing-svg:fill-current containing-svg:w-[1em]"
                  >
                    <svg
                      viewBox="0 0 16 16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M8 7C9.933 7 11.5 5.433 11.5 3.5C11.5 1.567 9.933 0 8 0C6.067 0 4.5 1.567 4.5 3.5C4.5 5.433 6.067 7 8 7Z"
                        fill="currentColor"
                      />
                      <path
                        d="M8 8.5C6.14413 8.50212 4.36489 9.2403 3.05259 10.5526C1.7403 11.8649 1.00212 13.6441 1 15.5C1 15.6326 1.05268 15.7598 1.14645 15.8536C1.24021 15.9473 1.36739 16 1.5 16H14.5C14.6326 16 14.7598 15.9473 14.8536 15.8536C14.9473 15.7598 15 15.6326 15 15.5C14.9979 13.6441 14.2597 11.8649 12.9474 10.5526C11.6351 9.2403 9.85587 8.50212 8 8.5Z"
                        fill="currentColor"
                      />
                    </svg>
                  </span>
                  peter
                </span>
              </a>
              <span
                class="flex items-center leading-none rounded-full overflow-hidden text-ellipsis mb-[-0.2em] pb-[0.2em] text-sm gap-1 px-2 py-1 text-slate-500 hover:text-slate-600 hover:bg-slate-100  group transition-all duration-300"
              >
                <span
                  class="inline-flex text-[11px] containing-svg:fill-current containing-svg:w-[1em]"
                >
                  <svg
                    viewBox="0 0 16 16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M14 3H13V1C13 0.734784 12.8946 0.48043 12.7071 0.292893C12.5196 0.105357 12.2652 0 12 0C11.7348 0 11.4804 0.105357 11.2929 0.292893C11.1054 0.48043 11 0.734784 11 1V3H5V1C5 0.734784 4.89464 0.48043 4.70711 0.292893C4.51957 0.105357 4.26522 0 4 0C3.73478 0 3.48043 0.105357 3.29289 0.292893C3.10536 0.48043 3 0.734784 3 1V3H2C1.46957 3 0.960859 3.21071 0.585786 3.58579C0.210714 3.96086 0 4.46957 0 5L0 14C0 14.5304 0.210714 15.0391 0.585786 15.4142C0.960859 15.7893 1.46957 16 2 16H14C14.5304 16 15.0391 15.7893 15.4142 15.4142C15.7893 15.0391 16 14.5304 16 14V5C16 4.46957 15.7893 3.96086 15.4142 3.58579C15.0391 3.21071 14.5304 3 14 3ZM2 14V7H14V14H2Z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
                <time
                  datetime="2023-05-03T04:55:45.155Z"
                  title="03.05.2023 06:55"
                >
                   vor 13 Tagen 
                </time>
              </span>
            </div>
          </div>
        </div>
        <div
          class="block font-medium leading-normal text-l [&>p]:mb-2 [&>p:last-child]:mb-0 [&>p>a]:text-violet-600 [&>p>a:hover]:text-violet-400 [&>p]:break-words"
        >
          <p>
            Hier ist noch ein 
            <a
              href="/tag/hashtag"
            >
              #hashtag
            </a>
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
